libraries/AP_OSD/AP_OSD_Backend.cpp

```diff
@@ -24,32 +24,57 @@ constexpr uint8_t AP_OSD_Backend::symbols[AP_OSD_NUM_SYMBOLS];
 #define SYM_DIG_OFS_1 0x90
 #define SYM_DIG_OFS_2 0xA0
 
+
+uint8_t AP_OSD_Backend::convert_to_decimal_packed_characters(char* buff, uint8_t size) {
+#if OSD_ENABLED
+    // use packed decimal characters based on fiam idea implemented in inav osd
+    // search the decimal separator with a bound and always terminate the string
+    char* p = (char*)memchr(&buff[1],'.',size-1);
+    if (p && isdigit(p[1]) && isdigit(p[-1])) {
+        // remove the decimal separator and replace the digit before and after
+        p[-1] += SYM_DIG_OFS_1;
+        p[1] += SYM_DIG_OFS_2;
+        // shift anything after p[1] 1 character to the left
+        const char* move_start = p+1;
+        const uint8_t move_size = size-(move_start-buff);
+        memmove(p, move_start, move_size);
+        p[move_size] = 0x0; // terminate
+        return size-1;
+    }
+#endif
+    return size;
+}
+
+uint8_t AP_OSD_Backend::format_string_for_osd(char* buff, uint8_t size, bool decimal_packed, const char *fmt, va_list ap)
+{
+    int res = 0;
+#if OSD_ENABLED
+    // note: vsnprintf() always terminates the string
+    res = hal.util->vsnprintf(buff, size, fmt, ap);
+    res = MIN(res, size);
+    if (res > 0 && decimal_packed) {
+        // note: convert_to_decimal_packed_characters() always terminates the string
+        res = convert_to_decimal_packed_characters(buff, res);
+    }
+#endif
+    return res;
+}
+
 void AP_OSD_Backend::write(uint8_t x, uint8_t y, bool blink, const char *fmt, ...)
 {
 #if OSD_ENABLED
     if (blink && (blink_phase < 2)) {
         return;
     }
-    char buff[32+1]; // +1 for snprintf null-termination
     va_list ap;
     va_start(ap, fmt);
-    int res = hal.util->vsnprintf(buff, sizeof(buff), fmt, ap);
-    res = MIN(res, int(sizeof(buff)));
-    if (res > 0 && check_option(AP_OSD::OPTION_DECIMAL_PACK)) {
-        // automatically use packed decimal characters
-        // based on fiam idea implemented in inav osd
-        char *p = strchr(&buff[1],'.');
-        if (p && isdigit(p[1]) && isdigit(p[-1])) {
-            p[-1] += SYM_DIG_OFS_1;
-            p[1] += SYM_DIG_OFS_2;
-            memmove(p, p+1, strlen(p+1)+1);
-            res--;
-        }
-    }
-    if (res < int(sizeof(buff))-1) {
-        write(x, y, buff);
-    }
+    char buff[32+1]; // +1 for null-termination
+    const uint8_t size = sizeof(buff);
+    // note: format_string_for_osd() always terminates the string
+    IGNORE_RETURN(format_string_for_osd(buff, size, check_option(AP_OSD::OPTION_DECIMAL_PACK), fmt, ap));
```

Comment:
If the only caller throws away the result, maybe just make it void.  Given the null-terminated guarantee a caller can just `strlen` it anyway.

Reply:
ok makes sense, will do